### PR TITLE
Adding a class to store and retrieve rules associated with an api

### PR DESCRIPTION
**Pull request by chkeita**
This PR introduce a new class to store and retrieve rules associated with an api endpoint.
This code will be used as part of the work to enable access control at the api level(#1074 )

### CODE REVIEW

File: src/api-service/__app__/onefuzzlib/request_access.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self.message = message
+
+
+class RequestAccess:
+    """
+    Stores the rules associated with a the request paths
+    """
+
+    class Rules:
+        allowed_groups_ids: List[UUID]
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/request_access.py:20]: User and group IDs are stored in the same list ?

File: src/api-service/__app__/onefuzzlib/request_access.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    root: Node
+
+    def __init__(self) -> None:
+        self.root = RequestAccess.Node()
+
+    def __add_url__(self, methods: List[str], path: str, rules: Rules) -> None:
+        methods = list(map(lambda m: m.upper(), methods))
+
+        segments = path.split("/")
+        if len(segments) == 0:
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/request_access.py:45]: if length is 0, does it mean that we apply rules to all APIs ?

File: src/api-service/__app__/onefuzzlib/request_access.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            self.rules = {}
+            self.children = {}
+            pass
+
+    root: Node
+
+    def __init__(self) -> None:
+        self.root = RequestAccess.Node()
+
+    def __add_url__(self, methods: List[str], path: str, rules: Rules) -> None:
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/request_access.py:41]: could you add unit tests for this as well ?

File: src/api-service/tests/test_request_access.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+import uuid
+
+from __app__.onefuzzlib.request_access import (
+    ApiAccessRule,
+    RequestAccess,
+    RuleConflictError,
+)
+
+
+class TestRequestAccess(unittest.TestCase):
```
Comment (stishkin) [inline src/api-service/tests/test_request_access.py:11]: missing test for "empty endpoint"

File: src/api-service/__app__/onefuzzlib/request_access.py (hunk 2 above)
Comment (chkeita) [reply src/api-service/__app__/onefuzzlib/request_access.py:45]: no it mens there are no rules

File: src/api-service/__app__/onefuzzlib/request_access.py (hunk 1 above)
Comment (chkeita) [reply src/api-service/__app__/onefuzzlib/request_access.py:20]: yes, might differentiate depending on the usage but that will be done in the PR that uses this class #1413

File: src/api-service/__app__/onefuzzlib/request_access.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        children: Dict[str, "RequestAccess.Node"]
+
+        def __init__(self) -> None:
+            self.rules = {}
+            self.children = {}
+            pass
+
+    root: Node
+
+    def __init__(self) -> None:
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/request_access.py:38]: Suggestion:
add print method to help visualize the tree structure

**Review by stishkin — APPROVED** [review]
